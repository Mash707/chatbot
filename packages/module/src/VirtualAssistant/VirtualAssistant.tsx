import React, { KeyboardEventHandler } from 'react';
import {
  Button,
  Card,
  CardBody,
  CardFooter,
  CardHeader,
  CardTitle,
  InputGroup,
  InputGroupText,
  TextArea
} from '@patternfly/react-core';
import { PaperPlaneIcon } from '@patternfly/react-icons';

export interface VirtualAssistantProps {
  /** Messages rendered within the assistant */
  children?: React.ReactNode;
  /** Header title for the assistant */
  title?: React.ReactNode;
  /** Input's placeholder for the assistant */
  inputPlaceholder?: string;
  /** Input's content */
  message?: string;
  /** Header actions of the assistant */
  actions?: React.ReactNode;
  /** Input's content change */
  onChangeMessage?: (event: React.ChangeEvent<HTMLTextAreaElement>, value: string) => void;
  /** Fire when clicking the Send (Plane) icon */
  onSendMessage?: (message: string) => void;
  /** Disables the text input */
  isInputDisabled?: boolean;
  /** Disables the send button */
  isSendButtonDisabled?: boolean;
}

export const VirtualAssistant: React.FunctionComponent<VirtualAssistantProps> = ({
  children,
  title = 'Virtual Assistant',
  inputPlaceholder = 'Type a message...',
  message = '',
  actions,
  onChangeMessage,
  onSendMessage,
  isInputDisabled = false,
  isSendButtonDisabled = false,
}: VirtualAssistantProps) => {

  const handleKeyPress: KeyboardEventHandler<HTMLTextAreaElement> = (event) => {
    if (event.key === 'Enter' || event.keyCode === 13) {
      if (!event.shiftKey) {
        if (message.trim() === '' || isSendButtonDisabled) {
          event.preventDefault();
        } else {
          onSendMessage && onSendMessage(message);
        }
      }
    }
  };

  return (
    <Card>
      <CardHeader actions={actions ? {
        actions
      } : undefined}>
        <CardTitle className="pf-v5-u-font-size-xl" data-test-id="assistant-title">
          {title}
        </CardTitle>
      </CardHeader>
      <CardBody >
        {children}
      </CardBody>
      <CardFooter>
        <InputGroup>
          <TextArea
            placeholder={inputPlaceholder}
            value={message}
            onChange={onChangeMessage}
            onKeyPress={handleKeyPress}
            type="text"
            aria-label="Assistant input"
            isDisabled={isInputDisabled}
            data-test-id="assistant-text-input"
          />
          <InputGroupText>
            <Button isDisabled={isSendButtonDisabled} data-test-id="assistant-send-button" aria-label="Virtual assistant's message" variant="plain" className="pf-v5-u-px-sm" onClick={onSendMessage ? () => {
              onSendMessage(message);
            } : undefined}>
              <PaperPlaneIcon />
            </Button>
          </InputGroupText>
        </InputGroup>
      </CardFooter>
    </Card>
  );
};

export default VirtualAssistant;
