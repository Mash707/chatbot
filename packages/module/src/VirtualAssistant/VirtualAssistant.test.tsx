import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import VirtualAssistant from './VirtualAssistant';

describe('VirtualAssistant', () => {
  it('should render assistant', () => {
    const { container } = render(<VirtualAssistant />);
    expect(container).toMatchSnapshot();
  });

  it('should render contents within the assistant', () => {
    render(
      <VirtualAssistant>
        <span>hello world</span>
      </VirtualAssistant>
    );
    expect(screen.getByText('hello world')).toBeTruthy();
  });

  it('should set custom title', () => {
    render(<VirtualAssistant title="I am a custom title" />);
    expect(screen.getByText('I am a custom title')).toBeTruthy();
  });

  it('should set custom input placeholder', () => {
    render(<VirtualAssistant inputPlaceholder="I am a custom placeholder" />);
    expect(screen.getByPlaceholderText('I am a custom placeholder')).toBeTruthy();
  });

  it('should set message', () => {
    render(<VirtualAssistant message="I am the message" />);
    expect(screen.getByText('I am the message')).toBeTruthy();
  });

  it('should listen to message changes', async () => {
    const listener = jest.fn();
    render(<VirtualAssistant onChangeMessage={listener} />);

    fireEvent.change(screen.getByRole('textbox'), {
      target: {
        value: 'hello textbox'
      }
    });

    expect(listener).toHaveBeenCalledTimes(1);
    expect(listener.mock.lastCall[1]).toBe('hello textbox');
  });

  it('should listen to send button press', async () => {
    const listener = jest.fn();
    render(<VirtualAssistant message={'I am the senate'} onSendMessage={listener} />);

    fireEvent.click(screen.getByRole('button'));

    expect(listener).toHaveBeenCalledTimes(1);
    expect(listener.mock.lastCall[0]).toBe('I am the senate');
  });

  it('should disable input', () => {
    render(<VirtualAssistant isInputDisabled={true} />);
    expect(screen.getByRole('textbox')).not.toBeEnabled();
  });

  it('should disable button', () => {
    render(<VirtualAssistant isSendButtonDisabled={true} />);
    expect(screen.getByRole('button')).not.toBeEnabled();
  });

  it('should trigger onSendMessage when pressing enter if there is a message', async () => {
    const user = userEvent.setup();
    const listener = jest.fn();
    render(<VirtualAssistant onSendMessage={listener} message="hello world" />);

    await user.type(screen.getByRole('textbox'), '[Enter]');

    expect(listener).toHaveBeenCalledTimes(1);
    expect(listener.mock.lastCall[0]).toBe('hello world');
  });

  it('should not trigger onSendMessage when pressing shift+enter if there is a message', async () => {
    const user = userEvent.setup();
    const listener = jest.fn();
    render(<VirtualAssistant onSendMessage={listener} message="hello world" />);

    await user.type(screen.getByRole('textbox'), '{Shift>}[Enter]{/Shift}');
    expect(listener).not.toHaveBeenCalled();
  });

  it('should not trigger onSendMessage when pressing enter if there is no text', async () => {
    const user = userEvent.setup();
    const listener = jest.fn();
    render(<VirtualAssistant onSendMessage={listener} message="" />);

    await user.type(screen.getByRole('textbox'), '[Enter]');
    expect(listener).not.toHaveBeenCalled();
  });

  it('should not trigger onSendMessage when pressing enter if there is only spaces / new lines and other empty content', async () => {
    const user = userEvent.setup();
    const message = '  \n\n\n  \t\t   \n\n\t\t  ';
    const listener = jest.fn();
    render(<VirtualAssistant onSendMessage={listener} message={message} />);

    await user.type(screen.getByRole('textbox'), '[Enter]');
    expect(listener).not.toHaveBeenCalled();
  });

  it('should not trigger onSendMessage when pressing enter if the send button is disabled', async () => {
    const user = userEvent.setup();
    const listener = jest.fn();
    render(<VirtualAssistant onSendMessage={listener} isSendButtonDisabled message="hello world" />);

    await user.type(screen.getByRole('textbox'), '[Enter]');
    expect(listener).not.toHaveBeenCalled();
  });
});
