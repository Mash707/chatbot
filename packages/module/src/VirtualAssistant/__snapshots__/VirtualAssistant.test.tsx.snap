// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VirtualAssistant should render assistant 1`] = `
<div>
  <div
    class="pf-v6-c-card"
    data-ouia-component-id="OUIA-Generated-Card-1"
    data-ouia-component-type="PF6/Card"
    data-ouia-safe="true"
    id=""
  >
    <div
      class="pf-v6-c-card__header"
    >
      <div
        class="pf-v6-c-card__header-main"
      >
        <div
          class="pf-v6-c-card__title"
        >
          <div
            class="pf-v6-c-card__title-text pf-v6-u-font-size-xl"
            data-test-id="assistant-title"
          >
            Virtual Assistant
          </div>
        </div>
      </div>
    </div>
    <div
      class="pf-v6-c-card__body"
    />
    <div
      class="pf-v6-c-card__footer"
    >
      <div
        class="pf-v6-c-input-group"
      >
        <span
          class="pf-v6-c-form-control pf-m-resize-both"
        >
          <textarea
            aria-invalid="false"
            aria-label="Assistant input"
            data-test-id="assistant-text-input"
            placeholder="Type a message..."
            type="text"
          />
        </span>
        <div
          class="pf-v6-c-input-group__item pf-m-box"
        >
          <span
            class="pf-v6-c-input-group__text"
          >
            <button
              aria-disabled="false"
              aria-label="Virtual assistant's message"
              class="pf-v6-c-button pf-m-plain pf-v6-u-px-sm"
              data-ouia-component-id="OUIA-Generated-Button-plain-1"
              data-ouia-component-type="PF6/Button"
              data-ouia-safe="true"
              data-test-id="assistant-send-button"
              type="button"
            >
              <span
                class="pf-v6-c-button__text"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v6-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 512 512"
                  width="1em"
                >
                  <path
                    d="M476 3.2L12.5 270.6c-18.1 10.4-15.8 35.6 2.2 43.2L121 358.4l287.3-253.2c5.5-4.9 13.3 2.6 8.6 8.3L176 407v80.5c0 23.6 28.5 32.9 42.5 15.8L282 426l124.6 52.2c14.2 6 30.4-2.9 33-18.2l72-432C515 7.8 493.3-6.8 476 3.2z"
                  />
                </svg>
              </span>
            </button>
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
